File: 1. array reduce/reduce.js
Array.prototype.my_reduce = function(f, acc) {
	if(!!this.length) acc = acc ? acc += this[0] : this[0];
	for(let i = 1; i < this.length; i++) acc = f(acc, this[i], i, this);
	return acc;
}

console.log([1, 2, 3, 4].my_reduce((acc, n) => acc + n));
console.log([1, 2, 3, 4].my_reduce((acc, n) => acc + n, 0));
console.log([].my_reduce((acc, n) => acc + n, 0));

File: 1. array reduce/reduce.rb
class Array
  def my_reduce(*args, &b)
    acc, m = args

    if block_given?
      each(&->(n) { acc = b.call(acc, n) })
    else
      each(&->(n) { acc = m(acc, n) })
    end

    acc
  end
end

pp [1, 2, 3].my_reduce(0) { |acc, n| acc + n }
pp [1, 2, 3].my_reduce 0, &->(acc, n) { acc + n }

pp [1, 2, 3].reduce(0) { |acc, n| acc + n }
pp [1, 2, 3].reduce 0, &->(acc, n) { acc + n }
